File: apps/api/src/users/utils/handle.validator.ts
export const HANDLE_MIN_LENGTH = 3;
export const HANDLE_MAX_LENGTH = 20;

const ALLOWED = /^[a-z0-9_]+$/;
const STARTS_WITH_ALNUM = /^[a-z0-9]/;

export const RESERVED_HANDLES = new Set([
  'admin', 'smwhr', 'support', 'help', 'official', 'staff', 'team',
  'api', 'root', 'test', 'demo', 'example', 'user', 'me',
]);

export function normalizeHandle(raw: string): string {
  let v = raw.trim().toLowerCase();
  if (v.startsWith('@')) v = v.slice(1);
  return v.replace(/\s/g, '');
}

export function validateHandle(raw: string): { ok: true } | { ok: false; reason: string } {
  const h = normalizeHandle(raw);
  if (h.length === 0) return { ok: false, reason: 'Handle is required' };
  if (h.length < HANDLE_MIN_LENGTH) {
    return { ok: false, reason: `Handle must be at least ${HANDLE_MIN_LENGTH} characters` };
  }
  if (h.length > HANDLE_MAX_LENGTH) {
    return { ok: false, reason: `Handle must be at most ${HANDLE_MAX_LENGTH} characters` };
  }
  if (!STARTS_WITH_ALNUM.test(h)) {
    return { ok: false, reason: 'Handle must start with a letter or digit' };
  }
  if (!ALLOWED.test(h)) {
    return { ok: false, reason: 'Handle may only contain a-z, 0-9 and underscore' };
  }
  if (RESERVED_HANDLES.has(h)) {
    return { ok: false, reason: 'Handle is reserved' };
  }
  return { ok: true };
}

File: apps/mobile/lib/shared/utils/handle_validator.dart
/// Pure validation for smwhr handles.
///
/// Rules (mirrors the backend that lands in Phase 2 — keep in sync):
/// - Length: 3..20 characters.
/// - Allowed: `a-z`, `0-9`, `_`. Lowercase only (we autolower in the field).
/// - Must start with a letter or digit (no leading underscore).
/// - Cannot be a reserved name (admin / smwhr / etc.) — owned by
///   `mock_users.dart`'s reservedHandles list.
///
/// Returns null when valid; otherwise a human-readable error.
class HandleValidator {
  HandleValidator._();

  static const int minLength = 3;
  static const int maxLength = 20;

  static final RegExp _allowed = RegExp(r'^[a-z0-9_]+$');
  static final RegExp _startsWithAlnum = RegExp(r'^[a-z0-9]');

  /// Cheap synchronous checks. Returns null when the handle passes the
  /// local rules; the live "is it taken" check happens via repository.
  static String? localError(String raw) {
    final h = raw.trim().toLowerCase();
    if (h.isEmpty) return null; // don't surface errors for empty input
    if (h.length < minLength) {
      return 'Mínimo $minLength caracteres.';
    }
    if (h.length > maxLength) {
      return 'Máximo $maxLength caracteres.';
    }
    if (!_startsWithAlnum.hasMatch(h)) {
      return 'Empieza con una letra o número.';
    }
    if (!_allowed.hasMatch(h)) {
      return 'Solo letras minúsculas, números y guion bajo.';
    }
    return null;
  }

  /// Normalises user input on the fly: lowercase, strip leading "@", strip
  /// spaces. Used by the text field's `onChanged` so the value the
  /// repository sees is always canonical.
  static String normalize(String raw) {
    var v = raw.trim().toLowerCase();
    if (v.startsWith('@')) v = v.substring(1);
    return v.replaceAll(' ', '');
  }
}
